### ISSUE TITLE
No way to set Truffle network config options for Truffle Test's or Truffle Develop's internal Ganache network

### ISSUE THREAD

**Issue by haltman-at** [trufflesuite/truffle#4327]
## Issue

Ordinarily, in Truffle, it is possible to set per-network configuration options in one's Truffle config.  However, as best I can tell, there is no way to do this for the ephemeral networks created by Truffle Test and Truffle Develop.  (The reason is different for each, so arguably this should be two issues, but I'm grouping them together.)  The Truffle Test issue has come up twice recently in other discussions (#3780, #4319) so I'm making an issue for it.

Let's start with Truffle Develop, as that's the easier case.  In the [`Environment.develop` function](https://github.com/trufflesuite/truffle/blob/develop/packages/environment/environment.js#L73), you can see that `config.networks[network]` is overwritten with a custom, barebones, network config.  Presumably, the new network config should keep other existing config options in place, and only overwrite `network_id`, `provider`, and other ways of specifying a provider; leaving in place such things as `gas`, `disableConfirmationListener`, etc.  This problem actually also applies to Truffle Test, as it also uses `Environment.develop`.

But troublesome problem with Truffle Test, and the one people have been asking about, is a different one.  Let's look at the [current behavior](https://github.com/trufflesuite/truffle/blob/develop/packages/core/lib/commands/test/index.js#L174).  Currently (a few lines above where I linked), if no network is specified, it will default to `"development"` or `"test"` (depending on if a config for `"development"` has been defined), and then (at the point where I linked), it will start up an internal Ganache (ordinarily called `"test"`) precisely if... no config is defined for the current network (whether explicitly selected or defaulted to).

This makes it impossible to configure this internal network.  A user might think, oh, well, the internal network is called `"test"`, so to configure it I'll add a config for the network `"test"`.  But as soon as they do so, Truffle will now say, oh, well since you have a config for `"test"`, you must have been intending to connect to some external network called `"test"`, and we won't start up Ganache for you.  That's not what the user wanted!  And people have been asking about this.  But right now there is no way to configure that internal network without Truffle concluding that want an external network instead.

It's not entirely clear what the solution to this is (without breaking backwards compatibility).  If we could break backwards compatibility, I'd just say that we should always start up a Ganache when the current network is `"test"`, regardless of whether a config exists or not, but I expect that would probably break what a lot of people are doing.  So, some other solution should be found to avoid breaking changes.  Perhaps we could try connecting to the network, and starting up Ganache if it doesn't exist?  Or there could be an option one could pass to force Truffle Test to start Ganache?  (It's annoying, though, because the people asking about this are likely just running `truffle test` and don't want to have to deal with additional options.  That said, that hasn't always been true of the people who have about this, so maybe I'm just wrong about that.)  Or maybe we could just set aside a new network name for the internal test network, like `"test-internal"` or something, and we could change it to be called that instead of `"test"`, since changing the name of the internal test network shouldn't break anything?  That might be the simplest solution, actually.

Anyway, one way or another we should come up with a fix for this.

## Environment

* Truffle version (`truffle version`): 5.4.11

**Comment by fainashalts** [comment]
Update: we will set a meeting to discuss the plan for this, likely after the holidays.

**Comment by gnidan** [comment]
Seems like path forward is that we call the unconfigurability of `"test"` a bug (and thus not a breaking change)

Group consensus (thanks @kevinweaver @eggplantzzz @haltman-at @fainashalts @cds-amal @tenthirtyone @leeftk for the discussion!) is that we should define detection criteria (such as `host`/`port`/`url`/`provider`), where the presence of at least one of these fields would indicate to Truffle that the network is unmanaged.

(Seems like there's some trickiness with `host`/`port`, if we want to allow these fields as configurable for managed networks in the future, but we won't worry about that for now.)

Thanks @haltman-at for volunteering to take a stab at this!

**Comment by haltman-at** [comment]
Yeah I'll note that I suggested at the meeting that we could ignore `port` and just look for `host`, `url`, or `provider`. :)

**Comment by lsqproduction** [comment]
Fixed in release v5.5.3

**Comment by Miladinho** [comment]
Hello, I am having this same issue, seems like it has been fixed but is there documentation for this? How do I keep my "networks" block available in the config but force "truffle test" to use its own internal ganache when it spins up?

**Comment by cds-amal** [comment]
@Miladinho do you mean you want a long-running ganache instance to be used for all your CI tests? This can be accomplished by:
- starting a ganache instance at the port you'd like
  ```ganache --port <my-port-number>```
- configuring a `development` network
  ```javascript
    networks: {
      development: {
       host: "127.0.0.1",     
       port: <my-port-number>,            
       network_id: "*",    
      },
    }
    // etc...
  ```

NOTE: You could also define a `test` network, however Truffle will use `development` if it also exists.

**Comment by Miladinho** [comment]
@cds-amal 

Not quite. I'm curious to see how i can get `truffle test` to default to its own ganache instant that it spins up. If I have a "networks" block defined in the config, it will force me to determine a network and uses the "development". If I do not have a "networks" block defined, running `truffle test` will work by using its own ganache instance i've read. Regardless of how it does it (let's not focus if i understood how this happens), it will run the tests. How do I get it to do this behavior and use its own test blockchain, instead of tryign to use a network i've defined in the config.

**Comment by haltman-at** [comment]
My understanding is that, as long as you don't have a `development` network, it will do this as long as no network is specified at the command line (or if `test`is the specified network).  I think if you do have a `development` network defined though it will default to that instead.

**Comment by Miladinho** [comment]
@haltman-at yes you're right, thanks. It would be nice to have that in documentation too. But also, do you think it is easier to have it as an option to force the internal/default ganache? Like `truffle test --default-ganache`

I also seem to have a problem with some RPC commands when I run with the internal vs using a self-defined network (ganache running locally). Basically, I have 2 tests failing when I don't use my own network. If you have any idea at the high level why that could happen I'd appreciate your advice. Thanks again

**Comment by haltman-at** [comment]
> But also, do you think it is easier to have it as an option to force the internal/default ganache? Like truffle test --default-ganache

Well, you could file an issue asking for this! :)

> I also seem to have a problem with some RPC commands when I run with the internal vs using a self-defined network (ganache running locally). Basically, I have 2 tests failing when I don't use my own network. If you have any idea at the high level why that could happen I'd appreciate your advice. Thanks again

No idea, sorry. I'd suggest either asking support or filing issue (as appropriate).
